libass/ass_shaper.c

```diff
@@ -451,54 +455,25 @@ get_contour_point(hb_font_t *font, void *font_data, hb_codepoint_t glyph,
 static hb_font_t *get_hb_font(ASS_Shaper *shaper, GlyphInfo *info)
 {
     ASS_Font *font = info->font;
-    hb_font_t **hb_fonts;
-
-    if (!font->shaper_priv)
-        font->shaper_priv = calloc(sizeof(ASS_ShaperFontData), 1);
-    if (!font->shaper_priv)
+    hb_font_t *hb_font = font->hb_fonts[info->face_index];
+    if (!hb_font)
         return NULL;
 
-    hb_fonts = font->shaper_priv->fonts;
-    if (!hb_fonts[info->face_index]) {
-        FT_Face face = font->faces[info->face_index];
-        hb_face_t *hb_face = hb_face_create_for_tables(get_reference_table, face, NULL);
-        if (!hb_face)
-            return NULL;
-        hb_face_set_index(hb_face, face->face_index);
-        hb_face_set_upem(hb_face, face->units_per_EM);
-
-        hb_font_t *hb_font = hb_fonts[info->face_index] = hb_font_create(hb_face);
-        hb_face_destroy(hb_face);
-        if (!hb_font)
-            return NULL;
-
-        hb_font_set_scale(hb_font,
-            (int)(((uint64_t)face->size->metrics.x_scale * face->units_per_EM + (1<<15)) >> 16),
-            (int)(((uint64_t)face->size->metrics.y_scale * face->units_per_EM + (1<<15)) >> 16));
-
-        // set up cached metrics access
-        struct ass_shaper_metrics_data *metrics =
-            font->shaper_priv->metrics_data[info->face_index] =
-                calloc(sizeof(struct ass_shaper_metrics_data), 1);
-        if (!metrics)
-            return NULL;
-        metrics->metrics_cache = shaper->metrics_cache;
-        metrics->vertical = info->font->desc.vertical;
-
-        hb_font_set_funcs(hb_font, shaper->font_funcs, metrics, NULL);
-    }
-
+    // set up cached metrics access
+    struct ass_shaper_metrics_data *metrics = calloc(sizeof(struct ass_shaper_metrics_data), 1);
+    if (!metrics)
+        return NULL;
     ass_face_set_size(font->faces[info->face_index], info->font_size);
-    update_hb_size(hb_fonts[info->face_index], font->faces[info->face_index]);
-
-    // update hash key for cached metrics
-    struct ass_shaper_metrics_data *metrics =
-        font->shaper_priv->metrics_data[info->face_index];
+    update_hb_size(hb_font, font->faces[info->face_index]);
+    metrics->metrics_cache = shaper->metrics_cache;
     metrics->hash_key.font = info->font;
     metrics->hash_key.face_index = info->face_index;
     metrics->hash_key.size = info->font_size;
+    metrics->vertical = info->font->desc.vertical;
+
+    hb_font_set_funcs(hb_font, shaper->font_funcs, metrics, free);
```

Comment:
Is it OK to take a shared `hb_font` from the `ASS_Font` and to apply a unique instance of `ass_shaper_metrics_data` to it via `hb_font_set_funcs`? Will this not interfere with threading?

Reply:
See `ass_shaper: don't mutate the shared hb_font_t during shaping` in #636 (currently 5ad73525b8834892cf958386c8dc6bc04915d97d); I use `hb_font_create_sub_font` to create a local `hb_font_t` that references the shared one, but has its own funcs font-data.

…though while writing this, I realized I could just cache the sub-font directly and save a decent bit of time for negligible memory cost, so I'm doing that.